base_layer/wallet/src/storage/sqlite_db/wallet.rs

```diff
@@ -472,99 +451,108 @@ fn get_db_encryption(
 
 fn get_cipher_for_db_encryption(
     passphrase: SafePassword,
-    passphrase_hash: Option<String>,
-    encryption_salt: Option<String>,
-    secret_seed: &Option<String>,
+    secondary_key_version: Option<String>,
+    secondary_key_salt: Option<String>,
+    encrypted_main_key: Option<String>,
     conn: &SqliteConnection,
 ) -> Result<XChaCha20Poly1305, WalletStorageError> {
-    let encryption_salt = match (passphrase_hash, encryption_salt) {
-        (None, None) => {
-            // Use the recommended OWASP parameters, which are not the default:
-            // https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#argon2id
-
-            // These are the parameters for the passphrase hash
-            let params_passphrase = argon2::Params::new(
-                46 * 1024, // m-cost: 46 MiB, converted to KiB
-                1,         // t-cost
-                1,         // p-cost
-                None,      // output length: default is fine for this use
-            )
-            .map_err(|e| WalletStorageError::AeadError(e.to_string()))?;
-
-            // Hash the passphrase to a PHC string for later verification
-            let passphrase_salt = SaltString::generate(&mut OsRng);
-            let passphrase_hash = argon2::Argon2::default()
-                .hash_password_customized(
+    // Set up the PBKDF used for secondary key derivation
+    // These parameters will likely change in the future, so we version them
+    // https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#argon2id
+    let pbkdf_params_version = 1u8;
+    let pbkdf_algorithm = argon2::Algorithm::Argon2id;
+    let pbkdf_version = argon2::Version::V0x13;
+    let pbkdf_params = argon2::Params::new(
+        46 * 1024,
+        1,
+        1,
+        Some(size_of::<Key>()), // for `XChaCha20-Poly1305`
+    )
+    .map_err(|e| WalletStorageError::AeadError(e.to_string()))?;
```

Comment:
Maybe these can go in a separate vector containing `pbkdf` struct items, similar to how `ConsensusConstants` is handled.

Reply:
Yeah, there will eventually need to be a more unified way to handle multiple versions anyway, since the client will need to be able to gracefully handle old versions and update to new ones. We can then include a version query that either returns the correct parameters, or raises a `BadEncryptionVersion` error.

Reply:
Worth noting that an even more general approach may eventually be needed, since there's no guarantee that all future versions will use `Argon2` as the PBKDF. But for now I think restricting to `Argon2` parameters should be fine.

Reply:
In one of my comments, I raise precisely this question. How can we minimize the impact on the user experience, given any need to update PBKDF parameters (including, version).

Reply:
Updates can be done seamlessly on startup, with no user interaction:
- The user enters their passphrase on startup as usual.
- We derive the secondary key using the out-of-date version parameters.
- The main key is authenticated and decrypted using the secondary key.
- We derive a new secondary key using the new version parameters (and the same passphrase).
- The main key is re-encrypted using the new secondary key.
- The new encrypted main key, secondary key salt, and version identifier are stored in the database.
- Life goes on!